runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java

```diff
@@ -84,36 +95,44 @@ public void channelRead(ChannelHandlerContext ctx, Object m)
             return;
         }
 
-        CorfuPayloadMsg<HandshakeResponse> handshakeResponse;
+        UUID serverId = null;
+        String corfuVersion = null;
+
+        if(m instanceof ResponseMsg) {
+            ResponseMsg response = ((ResponseMsg) m);
+
+            if(response.getPayload().hasHandshakeResponse()) {
+                log.info("channelRead: Handshake Response received. Removing {} from pipeline.",
+                        READ_TIMEOUT_HANDLER);
+
+                HandshakeResponseMsg handshakeResponse = response.getPayload().getHandshakeResponse();
+                corfuVersion = handshakeResponse.getCorfuVersion();
+                serverId = getUUID(handshakeResponse.getServerId());
 
-        try {
-            handshakeResponse = (CorfuPayloadMsg<HandshakeResponse>) m;
-            log.info("channelRead: Handshake Response received. Removing {} from pipeline.",
-                    READ_TIMEOUT_HANDLER);
-            // Remove the handler from the pipeline. Also remove the reference of the context from
-            // the handler so that it does not disconnect the channel.
-            ctx.pipeline().remove(READ_TIMEOUT_HANDLER).handlerRemoved(ctx);
-        } catch (ClassCastException e) {
-            log.warn("channelRead: Non-handshake message received by handshake handler. " +
-                    "Send upstream only if handshake succeeded.", e);
-            if (this.handshakeState.completed()) {
-                // Only send upstream if handshake is complete.
-                super.channelRead(ctx, m);
+                // Remove the handler from the pipeline. Also remove the reference of the context from
+                // the handler so that it does not disconnect the channel.
+                ctx.pipeline().remove(READ_TIMEOUT_HANDLER).handlerRemoved(ctx);
             } else {
-                // Otherwise, drop message.
-                try {
-                    CorfuMsg msg = (CorfuMsg) m;
+                log.warn("channelRead: Non-Handshake Response received {}", response);
+                if (this.handshakeState.completed()) {
```

Comment:
Check not needed.

Reply:
@vjeko Is there a reason why this is checked multiple times in the original code? I was under the impression that the handshake state could change between the first check and the second, so this was needed to avoid message loss.